File: HOT100/8.二叉树/437.路径总和-iii.py
#
# @lc app=leetcode.cn id=437 lang=python3
#
# [437] 路径总和 III
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        cnt = 0
        hash = defaultdict(int)
        hash[0] = 1
        def dfs(n, val):
            if not n: return
            val += n.val
            nonlocal cnt
            cnt += hash[val - targetSum]
            hash[val] += 1
            dfs(n.left, val)
            dfs(n.right, val)
            hash[val] -= 1
        dfs(root, 0)
        return cnt
# @lc code=end

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        cnt = 0
        def dfs(n, sums):
            if not n : return
            nonlocal cnt
            if sums == targetSum: cnt += 1
            if n.left:
                sums += n.left.val
                dfs(n.left, sums)
                sums -= n.left.val
            if n.right:
                sums += n.right.val
                dfs(n.right, sums)
                sums -= n.right.val
            return
            
        if not root: return 0
        q = deque()
        q.append(root)
        while q:
            tmp = q.popleft()
            dfs(tmp, tmp.val)
            if tmp.left: q.append(tmp.left)
            if tmp.right: q.append(tmp.right)
        
        return cnt

File: HOT100/8.二叉树/437.路径总和-iii.cpp
/*
 * @lc app=leetcode.cn id=437 lang=cpp
 *
 * [437] 路径总和 III
 */
#include "tools.h"
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int pathSum(TreeNode* root, int targetSum) {
        unordered_map<long long, long long> hash;
        hash[0] = 1;
        auto dfs = [&](auto&& self, TreeNode* n, long long val) -> void
        {
            if (! n) return;
            val += n->val;
            res += hash[val - targetSum];
            hash[val] += 1;
            self(self, n->left, val);
            self(self, n->right, val);
            hash[val] -= 1;
        };
        dfs(dfs, root, 0);
        return res;
    }
};
// @lc code=end

